JDBC/OAuthSampleApp/src/main/java/OAuthSampleApp.java

```diff
@@ -10,91 +10,298 @@
+	private static String getParam(String paramName) {
+		return csProp.get(paramName);
+	}
+
+	// Add/Create Authentication record in database. Create User and grant the
+	// permissions for User
+	private static void SetUpDbForOAuth() throws Exception {
+		String connectionString = getConnectionString();
+		vConnection = DriverManager.getConnection(connectionString);
+		String CONNECTION_STRING = prop.getProperty("ConnectionString");
+		String USER = prop.getProperty("User");
+		String CLIENT_ID = prop.getProperty("ClientId");
+		String CLIENT_SECRET = prop.getProperty("ClientSecret");
+		String DISCOVERY_URL = prop.getProperty("DiscoveryUrl");
+		Statement st = vConnection.createStatement();
+		st.execute("DROP USER IF EXISTS " + USER + " CASCADE;");
+		st.execute("DROP AUTHENTICATION IF EXISTS v_oauth CASCADE;");
+		st.execute("CREATE AUTHENTICATION v_oauth METHOD 'oauth' LOCAL;");
+		st.execute("ALTER AUTHENTICATION v_oauth SET client_id= '" + CLIENT_ID + "';");
+		st.execute("ALTER AUTHENTICATION v_oauth SET client_secret= '" + CLIENT_SECRET + "';");
+		st.execute("ALTER AUTHENTICATION v_oauth SET discovery_url = '" + DISCOVERY_URL + "';");
+		st.execute("CREATE USER " + USER + ";");
+		st.execute("GRANT AUTHENTICATION v_oauth TO " + USER + ";");
+		st.execute("GRANT ALL ON SCHEMA PUBLIC TO " + USER + ";");
+		st.close();
+	}
+
+	// Dispose the authentication record from database
+	private static void TearDown() {
+		try {
+			Statement st = vConnection.createStatement();
+			String USER = prop.getProperty("User");
+			st.executeUpdate("DROP USER IF EXISTS " + USER + " CASCADE");
+			st.executeUpdate("DROP AUTHENTICATION IF EXISTS v_oauth CASCADE");
+			vConnection.close();
+		} catch (Exception e) {
+			e.printStackTrace();
+		}
+	}
+
+	// Connect to Database using access Token
+	private static Connection connectToDB(String accessToken) throws SQLException {
+		Properties jdbcOptions = new Properties();
+		jdbcOptions.put("oauthaccesstoken", accessToken);
+		return DriverManager.getConnection(
+				"jdbc:vertica://" + getParam("Host") + ":" + getParam("Port") + "/" + getParam("Database"),
+				jdbcOptions);
+	}
+
+	// Test connection using access token and test database query result
+	private static void ConnectToDatabase() throws SQLException {
+		int connAttemptCount = 0;
+		while (connAttemptCount <= 1) {
+			try {
+				String accessToken = System.getProperty(OAUTH_ACCESS_TOKEN_VAR_STRING);
+				if (null == accessToken || accessToken.isEmpty()) {
+					throw new Exception("Access Token is not available.");
+				} else {
+					System.out.println("Attempting to connect with OAuth access token");
+					Connection conn = connectToDB(accessToken);
+					ResultSet rs = executeQuery(conn);
+					printResults(rs);
+					System.out.println("Query Executed. Exiting.");
+					break;
+				}
+			} catch (Exception ex) {
+				if (connAttemptCount > 0) {
+					break;
+				}
+				try {
+					GetTokensByRefreshGrant();
```

Comment:
I'd put a message here saying that the access token is invalid/expired, trying to do refresh.

Reply:
-> added a message